File: app/src/main/java/ys/moire/presentation/ui/view_parts/type/Rectangle.java
package ys.moire.presentation.ui.view_parts.type;

import android.graphics.Path;
import android.graphics.PointF;

/**
 * rectangle class.
 */
public class Rectangle extends BaseType {

    public PointF leftTop;
    public PointF rightTop;
    public PointF rightBottom;
    public PointF leftBottom;

    public float topLength;
    public float bottomLength;

    public Path path;

    public Rectangle() {

    }

    public void init(final float x, final float y, final float height, final float topLength, final float bottomLength) {
        this.topLength = topLength;
        this.bottomLength = bottomLength;
        this.leftTop = new PointF(x - topLength/2, y - height);
        this.rightTop = new PointF(x + topLength/2, y - height);
        this.leftBottom = new PointF(x - bottomLength/2, y + height);
        this.rightBottom = new PointF(x + bottomLength/2, y + height);

        path  = new Path();
        setPath();
    }

    /**
     * scroll rectangle automatically.
     * @param dx dx
     */
    @Override
    void autoMove(final float dx) {
        leftTop.x += dx;
        rightTop.x += dx;
        rightBottom.x += dx;
        leftBottom.x += dx;

        setPath();
    }

    /**
     * move x.
     */
    public void moveX(final float centerX) {
        rightBottom.x = centerX+bottomLength/2;
        leftBottom.x = centerX-bottomLength/2;
        leftTop.x = centerX-topLength/2;
        rightTop.x = centerX+topLength/2;
        setPath();
    }

    /**
     * add touch value.
     * @param touchDx  dx
     * @param touchDy  dy
     */
    public void addTouchVal(final int touchDx, final int touchDy) {
        leftTop.x += touchDx;
        leftTop.y += touchDy;
        rightTop.x += touchDx;
        rightTop.y += touchDy;
        rightBottom.x += touchDx;
        rightBottom.y += touchDy;
        leftBottom.x += touchDx;
        leftBottom.y += touchDy;
        setPath();
    }

    private void setPath() {
        path.reset();
        path.moveTo(leftTop.x, leftTop.y);
        path.lineTo(rightTop.x, rightTop.y);
        path.lineTo(rightBottom.x, rightBottom.y);
        path.lineTo(leftBottom.x, leftBottom.y);
        path.lineTo(leftTop.x, leftTop.y);
    }

}

File: app/src/main/java/ys/moire/presentation/ui/view_parts/type/Rectangle.kt
package ys.moire.presentation.ui.view_parts.type

import android.graphics.Path
import android.graphics.PointF

/**
 * rectangle class.
 */
class Rectangle : BaseType() {

    lateinit var leftTop: PointF
    lateinit var rightTop: PointF
    lateinit var rightBottom: PointF
    lateinit var leftBottom: PointF

    var topLength: Float = 0f
    var bottomLength: Float = 0f

    lateinit var path: Path

    fun init(x: Float, y: Float, height: Float, topLength: Float, bottomLength: Float) {
        this.topLength = topLength
        this.bottomLength = bottomLength
        this.leftTop = PointF(x - topLength / 2, y - height)
        this.rightTop = PointF(x + topLength / 2, y - height)
        this.leftBottom = PointF(x - bottomLength / 2, y + height)
        this.rightBottom = PointF(x + bottomLength / 2, y + height)

        path = Path()
        setPath()
    }

    /**
     * scroll rectangle automatically.
     * @param dx dx
     */
    override fun autoMove(dx: Float) {
        leftTop.x += dx
        rightTop.x += dx
        rightBottom.x += dx
        leftBottom.x += dx

        setPath()
    }

    /**
     * move x.
     */
    fun moveX(centerX: Float) {
        rightBottom.x = centerX + bottomLength / 2
        leftBottom.x = centerX - bottomLength / 2
        leftTop.x = centerX - topLength / 2
        rightTop.x = centerX + topLength / 2
        setPath()
    }

    /**
     * add touch value.
     * @param touchDx  dx
     * *
     * @param touchDy  dy
     */
    public override fun addTouchVal(touchDx: Int, touchDy: Int) {
        leftTop.x += touchDx.toFloat()
        leftTop.y += touchDy.toFloat()
        rightTop.x += touchDx.toFloat()
        rightTop.y += touchDy.toFloat()
        rightBottom.x += touchDx.toFloat()
        rightBottom.y += touchDy.toFloat()
        leftBottom.x += touchDx.toFloat()
        leftBottom.y += touchDy.toFloat()
        setPath()
    }

    private fun setPath() {
        path.reset()
        path.moveTo(leftTop.x, leftTop.y)
        path.lineTo(rightTop.x, rightTop.y)
        path.lineTo(rightBottom.x, rightBottom.y)
        path.lineTo(leftBottom.x, leftBottom.y)
        path.lineTo(leftTop.x, leftTop.y)
    }

}
